src/daemon/daemon.cpp

```diff
@@ -2668,6 +2680,82 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::clone(const CloneRequest* request,
+                       grpc::ServerReaderWriterInterface<CloneReply, CloneRequest>* server,
+                       std::promise<grpc::Status>* status_promise)
+{
```

Comment:
Can we get rid of this first set of braces and remove one set on indentation. That way it matches the format of the rest of the multipass commands in the daemon.

Reply:
I think the right place to use the function try block is the class ctor data member initializer list. See this [article](https://mariusbancila.ro/blog/2019/03/13/little-known-cpp-function-try-block/) for details. We should not abuse it anywhere else even though it is semantically the same here.

Reply:
Interesting read, thanks :slightly_smiling_face: However, I think the first part of the article is a little misleading:

> Exceptions from constructors and destructors of objects declared in the global namespace are not caught with function-try-catch in main(). 

They are [not caught with a regular try-catch either](https://godbolt.org/z/TcG7jjW5G). So this is not a reason to prefer one over the other.

> Exceptions from destructors of static objects are not caught with function-try-catch in main(). 

Again, [a regular try-catch doesn't help here either](https://godbolt.org/z/3chrGGG8c). 

In both cases, the beginning/end of the object's lifetime is just outside of the boundaries of main. The corollary is just that global objects need to be used with care where exceptions are concerned (and destructors should not throw anyway).

> If a function (any function, not just main()) has a return type other than void and the function-try-catch does not have a return statement in the catch block then the behaviour is undefined. 

This is a good reason to prefer the regular try-catch. However, in this case, all of the functions are `void`. An argument can be made that the function-try is error prone if we start returning from those functions. On the other hand, there is the argument of readability (one level of indentation can result in more line breaks) and the one of consistency with the other methods.

Reply:
@ricab 
Thanks for reading and sharing the feedback. There are some uncommon use cases of regular try-catch vs function try-catch in the article. But the core message of the article is `"So what is then the real use of function try-catch? The answer is catching exceptions in constructors member initializer list"`. In a nutshell, using function try-catch when the regular one can not. Because the syntax of it is just not common, in a way, personally I think it is less readable because it is harder to capture and navigate on the scope of functions like `daemon::clone`, `daemon::snapshot`. For example, you can not jump from the beginning curly bracket of the function to the ending one because the scope is scattered. 

> This is a good reason to prefer the regular try-catch. However, in this case, all of the functions are void. 

I already mentioned that they are sematically the same in this particular scenario. It is just a little abusive of function try-catch. 

For consistency, yes, we already have that all over the place. I do not have a strong opinion on which one to choose, just want to raise the concern and make the team understand the nuisance of these two.

Reply:
I think you have good points, especially in your concerns for readability and navigability. However, I think that classifying constructors as the only valid use of function try is disputable. The language allows the way we are using it, it is well defined, so I am not convinced it is really abusive. And most of the arguments in the article you shared fail to support that "core message".

In the end, I guess it is kind of subjective which is the better approach. FWIW, I don't mind discussing this more broadly and possibly changing all those methods to a different scheme we agree with. So feel free to bring this up :slightly_smiling_face: 

For now, I would say that current practice in the other methods in that file is a good tie-breaker and we just stick with it. Does that sound reasonable to you?